core/common/src/unsafe/UnsafeBufferOperations.kt

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright 2010-2024 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE.txt file.
+ */
+
+package kotlinx.io.unsafe
+
+import kotlinx.io.*
+
+@UnsafeIoApi
+public object UnsafeBufferOperations {
```

Comment:
Does it really make sense to wrap these functions in an object?
Why not going top-level?

I understand it acts as a sort-of namespace, however it means it's an additional instance, and in K/JS each function call well be behind an `undefined` check as everything is done lazily.

Reply:
We're trying hard to ensure users understand that the API is unsafe and they REALLY need to use it.

The necessity to call it via `UnsafeBufferOperations` is yet another hint at the call site that the API is unsafe.

>  in K/JS each function call will be behind an undefined check as everything is done lazily.

Do you expect it to be a performance bottleneck for this particular API? 
(The following is a speculation, mostly) `undefined` check should be a single comparison followed by the jump, right?

Reply:
> We're trying hard to ensure users understand that the API is unsafe and they REALLY need to use it.

Reasonable! But wouldn't the unsafe opt-in alone on each function be enough?

> Do you expect it to be a performance bottleneck for this particular API?

If we take kotlinx-io in isolation, definitely not.  
What I worry about is the build-up of lazy initialization checks. Some are contributed by io, some by coroutines, some by serialization, some by other third party libraries, etc.

That's why I always ask if it's really necessary. I hope [namespacing in the language](https://youtrack.jetbrains.com/issue/KT-11968) will remove the necessity of using objects for that.

Reply:
> But wouldn't the unsafe opt-in alone on each function be enough?

One can opt it in for a file or the whole project. 😞

Reply:
But it's the same if a user uses static-like imports:

```kotlin
object MyObject {
  fun one() {}
  fun two() {}
}
```

```kotlin
import com.test.MyObject.one
import com.test.MyObject.two

fun example() {
  one()
  two()
}
```

So I wouldn't worry too much about it.

Reply:
> But it's the same if a user uses static-like imports

Sure, but users have to add a project/file-wide opt-in and write a static import. That lowers the chances of accidentally using unsafe API.

It's also worth mentioning that there's no goal of making that API convenient for everyday use. It's quite the opposite: regular kotlinx-io users are not expected to use unsafe API, and a clumsier API shape is one of the leverages to nudge them toward using other APIs.

Reply:
I'll have to see the code once it's all done, but the thing you have to keep in mind is code like this

```kotlin
while (condition) {
  UnsafeBufferOperations.someFun(...)
}
```

where there are thousand+ iterations possible. That's where the `undefined` check becomes a problem.

Reply:
https://github.com/fzhinkin/kjs-obj-recvr-overhead/blob/main/src/commonMain/kotlin/FunctionCallBenchmark.kt

https://jmh.morethan.io/?source=https://raw.githubusercontent.com/fzhinkin/kjs-obj-recvr-overhead/main/results/2024-06-10T13.32.24.647154309-kn-js.json

It is hard to argue about an `undefined` check's impact on the performance given the high jitter, but it seems like it does not exceed the measurement error.

Reply:
For others, see [this Slack post](https://kotlinlang.slack.com/archives/CKCAAV8VB/p1718034129228599).